### DISCUSSION TITLE
Consume Asynchronous REST API with dagster

### DISCUSSION THREAD

**Discussion by slamer59** [dagster-io/dagster#18401] [Q&A] [answered]
Hello, I have an external service in REST I want to cosume with dagster.
Two use case :
- event based : A post request is made with a callback url
- Queue based : A post is request. an id for the identificated task. A get request is made on this ID to check if the job is done.
How would you handle those kind of tasks ? I don't find any exemple.
For queue based, I guess a sensor is OK. but for the 1st one I don't know without creating some middleware.

**Comment by smackesey** [comment]
Hi, can you elaborate on "event-based" here? How does the service to which you're making requests relate to Dagster?

**Comment by slamer59** [comment]
Hi @smackesey 
a better name is Webhooks. 
A callback url is registered in the external service.  

If I call the service, an event is triggered when a long task is finished. 

For exemple : 
https://docs.wave.com/business?shell#webhooks

```

1. You create one or more webhook endpoints on your server to receive events from Wave.
2. You register the URLs of your webhook endpoints and choose which events you want to be notified about for each.
3. When an event occurs, Wave sends a POST request to the URLs that have registered for the event. The body of the request is an [Event](https://docs.wave.com/business?shell#events) object
4. Upon receiving an event notification, your webhook must verify that the event was sent by Wave by validating the Wave-Signature header.
5. Respond to the event with a successful status code (HTTP 2xx) before performing any business logic.
6. Perform business logic in response to the event.

```

**Reply by smackesey** [reply]
Hmm, I'm still not understanding what you want to do with Dagster. Can you describe the desired interaction between your external service and Dagster jobs/assets etc?

**Reply by slamer59** [reply]
I try to be schematic. 
It can be another setup but it is the way I see it. 

```mermaid
graph TD
  subgraph ExternalService
    A[Register Callback URL : URL1] 
    B[LongTask]
  end
  subgraph Dagster1
    F[Asset1]
    G[Asset2]
  end

  subgraph Dagster2
    C[Resolved Asset2 asset with LongTaskID]
    D[Asset3]
    E[Asset4]
  end

  
  G-->|1. POST request. Receive an LongTaskID|B
  F-->G
  B -->|2. LongTaskID Finished : send event to URL1 | C
  C -->D
  D -->E

```


The URL1 registration depends on the external service. 

We could store LongTaskID and create a sensor but it is not efficient as we don't know in advance how much time the LongTask runs (and not triggered by event). 

I hope it is clearer now :)

**Reply by slamer59** [reply]
The same in a sequence
```mermaid

sequenceDiagram
  participant ExternalService
  participant DagsterJob1
  participant DagsterJob2
  
  ExternalService ->> ExternalService: Register Callback URL : URL1


  activate DagsterJob1
  DagsterJob1 ->> ExternalService: POST request

  activate ExternalService
  ExternalService -->> DagsterJob1: Receive a LongTaskID

  deactivate DagsterJob1
  ExternalService -->> ExternalService: Running LongTaskID

  
  ExternalService -->> DagsterJob2: POST LongTaskID Finished

  activate DagsterJob2



  deactivate ExternalService
  DagsterJob2 ->> DagsterJob2: Resolve Asset2 in some way ?

  DagsterJob2 ->> DagsterJob2: Business Logic
  deactivate DagsterJob2

```

**Reply by smackesey** [reply, marked as answer]
OK let's see if I understand.

- You have two jobs in Dagster. Job 1 (J1) materializes Asset 1 (A1) and Asset 2 (A2). Job 2 (J2) materializes Asset 3 (A3) and Asset 4 (A4).
- A2 places a POST request to the External Service, including some callback URL. The service responds with a Task ID. We don't know how long the task will take to finish.
- The External Service sends a POST request to the callback URL when the submitted task is finished.
- You want to record a materialization for A2 and launch J2 upon receiving the POST to the callback URL.

If that's the case, here are some comments/recommendations:

- If A2 takes so long to compute (by the external service) that you don't want to keep a Dagster run going, then you shouldn't use a typical SDA for it, but rather an [External Asset](https://docs.dagster.io/concepts/assets/external-assets). You should submit your POST to the external service in an op. You will need to run some other server to listen for the callback, then you can use the `report_asset_materialization` endpoint of our REST API to report a materialization for A2 when you receive this callback.
- You can run an [asset sensor](https://docs.dagster.io/concepts/partitions-schedules-sensors/asset-sensors) to listen for materializations of A2 and launch J2 in response.

**Reply by slamer59** [reply]
Hi, thanks for your view.

I will look external assets as it fits the definition. 

For the sensor part, it will work but less efficient pattern compared to event based pattern. 

I guess I will need to trick and an intermediate endpoint which will convert API to a graphql dagster call.
